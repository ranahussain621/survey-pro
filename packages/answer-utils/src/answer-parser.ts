import { CURRENCY_SYMBOLS } from './consts'
import { Answer, FullNameValue, ServerSidePaymentValue } from '@heyform-inc/shared-types-enums'
import { helper } from '@heyform-inc/utils'
import Big from 'big.js'

function fileUpload(answer: Answer, livePreview = false) {
  if (livePreview) {
    return {
      filename: (answer as any).name
    }
  }

  if (helper.isURL(answer.value)) {
    return {
      filename: '',
      url: answer.value
    }
  }

  return {
    filename: answer.value.filename || '',
    url: answer.value.url
  }
}

function rating(answer: Answer): string {
  return answer.value
}

function singleChoice(answer: Answer): string {
  const choice = answer.properties!.choices?.find(row => row.id === answer.value)
  return choice ? choice.label : ''
}

function multipleChoice(answer: Answer): string {
  return (answer.properties?.choices?.filter(row => answer.value.value.includes(row.id)) || [])
    .map(row => row.label)
    .concat([answer.value.other])
    .filter(row => helper.isValid(row))
    .join(', ')
}

function fullName(answer: Answer): FullNameValue {
  return answer.value
}

function address(answer: Answer): string {
  return [
    answer.value.address1,
    ',',
    answer.value.address2,
    answer.value.city,
    ',',
    answer.value.state,
    ',',
    answer.value.country,
    answer.value.zip
  ]
    .filter(Boolean)
    .join(' ')
}

function legalTerms(answer: Answer): string {
  return helper.isTrue(answer.value) ? 'Yes' : 'No'
}

function dateRange(answer: Answer): string {
  return [answer.value.start, answer.value.end].filter(Boolean).join(' - ')
}

function inputTable(answer: Answer): string {
  const columns = answer.properties?.tableColumns

  if (helper.isValidArray(columns)) {
    const result: string[] = []

    answer.value.forEach((values: Record<string, string>) => {
      if (helper.isValid(values)) {
        const row = columns!.map(column => values[column.id]).join(', ')
        result.push(row)
      }
    })

    return result.join('\n')
  }

  return ''
}

function payment(answer: Answer): string {
  const value = answer.value as ServerSidePaymentValue
  const price = Big(value.amount).div(100).toFixed(2)
  let result = CURRENCY_SYMBOLS[value.currency] + price

  if (helper.isValid(value.paymentIntentId)) {
    result = `Succeeded ${result}`
  } else {
    result = `Incomplete ${result}`
  }

  return result
}

export default {
  fileUpload,
  rating,
  singleChoice,
  multipleChoice,
  fullName,
  address,
  legalTerms,
  dateRange,
  inputTable,
  payment
}
