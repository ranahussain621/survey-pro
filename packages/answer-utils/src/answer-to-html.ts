import parser from './answer-parser'
import { Answer, FieldKindEnum, QUESTION_FIELD_KINDS } from '@heyform-inc/shared-types-enums'

export function answersToHtml(answers: Answer[]): string {
  const html = answers
    .map(answer => {
      const value = parseHtmlAnswer(answer)

      return `
<li>
  <h3>${answer.title}</h3>
  <p>${value}</p>
</li>
`
    })
    .join('')

  return `<ol>${html}</ol>`
}

function parseHtmlAnswer(answer: Answer): string {
  let value = ''

  if (QUESTION_FIELD_KINDS.includes(answer.kind)) {
    switch (answer.kind) {
      case FieldKindEnum.FILE_UPLOAD:
        const file = parser.fileUpload(answer)
        value = `${file.filename} (${file.url})`
        break

      case FieldKindEnum.RATING:
      case FieldKindEnum.OPINION_SCALE:
        value = parser.rating(answer)
        break

      case FieldKindEnum.YES_NO:
        value = parser.singleChoice(answer)
        break

      case FieldKindEnum.MULTIPLE_CHOICE:
      case FieldKindEnum.PICTURE_CHOICE:
        value = parser.multipleChoice(answer)
        break

      case FieldKindEnum.FULL_NAME:
        const name = parser.fullName(answer)
        value = `${name.firstName} ${name.lastName}`
        break

      case FieldKindEnum.ADDRESS:
        value = parser.address(answer)
        break

      case FieldKindEnum.LEGAL_TERMS:
        value = parser.legalTerms(answer)
        break

      case FieldKindEnum.DATE_RANGE:
        value = parser.dateRange(answer)
        break

      case FieldKindEnum.INPUT_TABLE:
        value = parser.inputTable(answer)
        break

      case FieldKindEnum.PAYMENT:
        value = parser.payment(answer)
        break

      default:
        value = answer.value?.toString()
    }
  }

  return value
}
