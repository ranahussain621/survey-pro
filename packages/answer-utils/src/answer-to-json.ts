import parser from './answer-parser'
import { Answer, FieldKindEnum } from '@heyform-inc/shared-types-enums'

interface AnswersToJsonOptions {
  plain?: boolean
}

export function answersToJson(
  answers: Answer[],
  options?: AnswersToJsonOptions
): Record<string, any> {
  const result: Record<string, any> = {}

  answers.forEach(answer => {
    result[answer.id] = parseJsonAnswer(answer, options?.plain)
  })

  return result
}

function parseJsonAnswer(answer: Answer, plain = false): any {
  let value: any

  switch (answer.kind) {
    case FieldKindEnum.FILE_UPLOAD:
      value = parser.fileUpload(answer)
      if (plain) {
        value = `${value.filename} (${value.url})`
      }
      break

    case FieldKindEnum.RATING:
    case FieldKindEnum.OPINION_SCALE:
      value = parser.rating(answer)
      break

    case FieldKindEnum.YES_NO:
      value = parser.singleChoice(answer)
      break

    case FieldKindEnum.MULTIPLE_CHOICE:
    case FieldKindEnum.PICTURE_CHOICE:
      value = parser.multipleChoice(answer)
      break

    case FieldKindEnum.FULL_NAME:
      value = parser.fullName(answer)
      if (plain) {
        value = `${value.firstName} ${value.lastName}`
      }
      break

    case FieldKindEnum.ADDRESS:
      value = answer.value
      if (plain) {
        value = parser.address(answer)
      }
      break

    case FieldKindEnum.LEGAL_TERMS:
      value = parser.legalTerms(answer)
      break

    case FieldKindEnum.DATE_RANGE:
      value = answer.value
      if (plain) {
        value = parser.dateRange(answer)
      }
      break

    case FieldKindEnum.INPUT_TABLE:
      value = answer.value
      if (plain) {
        value = parser.inputTable(answer)
      }
      break

    case FieldKindEnum.PAYMENT:
      value = answer.value
      if (plain) {
        value = parser.payment(answer)
      }
      break

    default:
      value = answer.value?.toString()
  }

  return value
}
