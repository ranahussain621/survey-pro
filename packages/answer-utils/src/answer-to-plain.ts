import parser from './answer-parser'
import { Answer, FieldKindEnum } from '@heyform-inc/shared-types-enums'
import { helper } from '@heyform-inc/utils'

export function answersToPlain(answers: Answer[]): string {
  return answers
    .map(answer => {
      const value = parsePlainAnswer(answer)
      return `${answer.title}\n${value}`
    })
    .join('\n\n')
}

export function parsePlainAnswer(answer: Answer, livePreview = false): string {
  let value: string

  switch (answer.kind) {
    case FieldKindEnum.FILE_UPLOAD:
      const file = parser.fileUpload(answer, livePreview)
      value = livePreview ? file.filename : `${file.filename} (${file.url})`
      break

    case FieldKindEnum.RATING:
    case FieldKindEnum.OPINION_SCALE:
      value = parser.rating(answer)
      break

    case FieldKindEnum.YES_NO:
      value = parser.singleChoice(answer)
      break

    case FieldKindEnum.MULTIPLE_CHOICE:
    case FieldKindEnum.PICTURE_CHOICE:
      value = parser.multipleChoice(answer)
      break

    case FieldKindEnum.FULL_NAME:
      const name = parser.fullName(answer)
      value = `${name.firstName} ${name.lastName}`
      break

    case FieldKindEnum.ADDRESS:
      value = parser.address(answer)
      break

    case FieldKindEnum.LEGAL_TERMS:
      value = parser.legalTerms(answer)
      break

    case FieldKindEnum.DATE_RANGE:
      value = parser.dateRange(answer)
      break

    case FieldKindEnum.INPUT_TABLE:
      value = parser.inputTable(answer)
      break

    case FieldKindEnum.PAYMENT:
      value = parser.payment(answer)
      break

    default:
      value = answer.value?.toString()
  }

  if (livePreview && helper.isEmpty(value)) {
    value = '_____'
  }

  return value
}
