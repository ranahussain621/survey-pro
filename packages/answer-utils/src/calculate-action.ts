import { isNumber } from './helper'
import {
  AnswerValue,
  CalculateEnum,
  NumberCalculateAction,
  StringCalculateAction,
  Variable
} from '@heyform-inc/shared-types-enums'
import { helper } from '@heyform-inc/utils'

export function calculateAction(
  action: NumberCalculateAction | StringCalculateAction,
  parameters?: Variable[],
  data?: Record<string, string | number>,
  values?: Record<string, AnswerValue>
) {
  if (helper.isEmpty(parameters) || helper.isEmpty(data) || helper.isNil(data![action.variable])) {
    return data || {}
  }

  const variable = parameters!.find(v => v.id === action.variable)!
  let current = data![action.variable]

  switch (variable.kind) {
    case 'string':
      current = calculateString(current as string, action as StringCalculateAction, values)
      break

    case 'number':
      current = calculateNumber(current as number, action as NumberCalculateAction, values)
      break
  }

  return {
    ...data,
    [action.variable]: current
  }
}

function calculateString(
  value: string,
  action: StringCalculateAction,
  values?: Record<string, AnswerValue>
): string {
  let newValue = action.value as string

  if (action.ref) {
    newValue = values?.[action.ref] as string
  }

  if (!helper.isNil(newValue)) {
    switch (action.operator) {
      case CalculateEnum.ADDITION:
        return value + newValue!

      case CalculateEnum.ASSIGNMENT:
        return newValue!
    }
  }

  return value
}

function calculateNumber(
  value: number,
  action: NumberCalculateAction,
  values?: Record<string, AnswerValue>
): number {
  let newValue = action.value as number

  if (action.ref) {
    newValue = values?.[action.ref] as number
  }

  if (!helper.isNil(newValue) && isNumber(newValue!)) {
    switch (action.operator) {
      case CalculateEnum.ADDITION:
        return value + newValue!

      case CalculateEnum.SUBTRACTION:
        return value - newValue!

      case CalculateEnum.MULTIPLICATION:
        return value * newValue!

      case CalculateEnum.DIVISION:
        return newValue! !== 0 ? value / newValue! : value

      case CalculateEnum.ASSIGNMENT:
        return newValue!
    }
  }

  return value
}
