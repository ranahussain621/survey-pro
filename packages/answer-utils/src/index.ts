export * from './answer-to-html'
export * from './answer-to-api-object'
export * from './answer-to-plain'
export * from './answer-to-json'
export * from './field-values-to-answer'
export * from './fields-to-validate-rules'
export * from './validate'
export * from './html-utils'
export * from './flatten-fields'
export * from './validate-condition'
export * from './validate-payload'
export * from './apply-logic-to-fields'
export * from './consts'
export * from './helper'
