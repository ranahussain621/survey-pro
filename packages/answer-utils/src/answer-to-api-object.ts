import { parsePlainAnswer } from './answer-to-plain'
import { Answer } from '@heyform-inc/shared-types-enums'

export function answersToApiObject(answers: Answer[]): Record<string, any> {
  const result: Record<string, any> = {}

  answers.forEach(answer => {
    const key = `(ID: ${answer.id}) ${answer.title}`
    result[key] = parsePlainAnswer(answer)
  })

  return result
}
