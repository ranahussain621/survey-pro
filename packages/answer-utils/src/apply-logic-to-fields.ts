import { calculateAction } from './calculate-action'
import { validateRequiredField } from './validate'
import { validateCondition } from './validate-condition'
import {
  ActionEnum,
  AnswerValue,
  FieldKindEnum,
  FormField,
  Logic,
  NavigateAction,
  NumberCalculateAction,
  QUESTION_FIELD_KINDS,
  StringCalculateAction,
  Variable
} from '@heyform-inc/shared-types-enums'
import { helper } from '@heyform-inc/utils'

interface IFormField extends FormField {
  parent?: FormField
  isTouched?: boolean
}

export interface PickAndCalcFieldsResult {
  fields: IFormField[]
  variables: Record<string, AnswerValue>
}

function indexFields(fields: IFormField[]) {
  const parents = fields.filter(f => !f.parent && QUESTION_FIELD_KINDS.includes(f.kind))
  const children = fields.filter(f => f.parent && QUESTION_FIELD_KINDS.includes(f.kind))

  let index = 1
  const childrenIndexes: Record<string, number> = {}

  parents.forEach(f => {
    f.index = index++
  })

  children.forEach(f => {
    const parentId = f.parent!.id
    const parent = parents.find(p => p.id === parentId)

    if (!childrenIndexes[parentId]) {
      childrenIndexes[parentId] = 1
    }

    f.parent = parent
    f.index = childrenIndexes[parentId]++
  })
}

export function applyLogicToFields(
  fields?: IFormField[],
  logics?: Logic[],
  parameters?: Variable[],
  values?: Record<string, AnswerValue>
): PickAndCalcFieldsResult {
  const result: PickAndCalcFieldsResult = {
    fields: [],
    variables: {}
  }

  if (helper.isEmpty(fields)) {
    return result
  }

  if (helper.isValid(parameters)) {
    for (const variable of parameters!) {
      result.variables[variable.id] = variable.value
    }
  }

  if (helper.isEmpty(logics) || helper.isEmpty(values)) {
    // Add indexes to every field
    indexFields(fields!)

    return {
      ...result,
      fields: fields!.filter(f => f.kind !== FieldKindEnum.THANK_YOU)
    }
  }

  let index = 0

  while (index < fields!.length) {
    const field = fields![index]

    if (field.kind === FieldKindEnum.THANK_YOU) {
      break
    }

    let isNavigateValidated = false
    const logic = logics!.find(l => l.fieldId === field.id)

    // Add parent field if needed to present SubGroup fields
    if (field.parent && result.fields.findIndex(f => f.id === field.parent!.id) < 0) {
      result.fields.push(field.parent)
    }

    if (logic) {
      const { payloads } = logic
      const calculates = payloads.filter(p => p.action.kind === ActionEnum.CALCULATE)

      for (const calculate of calculates) {
        const { action, condition } = calculate
        const isValidated = validateCondition(field, condition, values)

        if (isValidated) {
          result.variables = calculateAction(
            action as NumberCalculateAction | StringCalculateAction,
            parameters,
            result.variables,
            values
          )
        }
      }

      const navigates = payloads.filter(p => p.action.kind === ActionEnum.NAVIGATE)

      // Find the first navigate which meets the jump rule
      for (const navigate of navigates) {
        const { action, condition } = navigate

        // Whether the jump logic in this field has been verified or not
        field.isTouched = validateCondition(field, condition, values)

        if (field.isTouched) {
          const jumpFieldId = (action as NavigateAction).fieldId
          const jumpIndex = fields!.findIndex(f => f.id === jumpFieldId)
          const isExists = !!result.fields.find(f => f.id === jumpFieldId)

          // Logic jump only works when sending the respondent forward in the form
          if (!isExists && jumpIndex > index) {
            index = jumpIndex
            isNavigateValidated = true

            result.fields.push(field)
            break
          }
        } else {
          field.isTouched = validateRequiredField(field, values)
        }
      }
    }

    if (!isNavigateValidated) {
      index += 1
      result.fields.push(field)
    }
  }

  // Add indexes to every field
  indexFields(result.fields)

  return result
}
