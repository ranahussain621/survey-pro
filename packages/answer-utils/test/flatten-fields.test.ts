import { flattenFields } from '../src'
import fields from './fixtures/fields.json'

test('should flatten fields', () => {
  expect(flattenFields(fields as any)).toMatchSnapshot()
})

test('should flatten fields with group', () => {
  const rawFields: any[] = [
    {
      properties: {
        fields: [
          {
            id: 'sub1',
            kind: 'multiple_choice'
          }
        ]
      },
      id: 'grp1',
      kind: 'group'
    },
    {
      id: 'txt1',
      kind: 'text'
    }
  ]

  expect(flattenFields(rawFields).map(f => f.id)).toStrictEqual(['sub1', 'txt1'])
  expect(flattenFields(rawFields, true).map(f => f.id)).toStrictEqual(['grp1', 'sub1', 'txt1'])
})
