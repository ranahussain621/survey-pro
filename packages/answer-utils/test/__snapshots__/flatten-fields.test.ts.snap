// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should flatten fields 1`] = `
Array [
  Object {
    "description": "Three Rings for Elven-kings under the sky,
Seven for the Dwarf-lords in their halls of stone,
Nine for Mortal Men doomed to die,

One for the Dark Lord on his dark throne
In the Land of Mordor where the shadows lie.

One Ring to rule them all, One Ring to find them,
One Ring to bring them all and in the darkness bind them
In the Land of Mordor where the shadows lie.",
    "id": "Z90w4JfKzc9c",
    "kind": "statement",
    "properties": Object {
      "choices": Array [],
    },
    "title": "",
    "validations": Object {
      "required": true,
    },
  },
  Object {
    "id": "5c8Rgj9et1r2",
    "kind": "text",
    "properties": Object {
      "allowMultiple": true,
    },
    "title": "Description",
    "validations": Object {
      "required": false,
    },
  },
  Object {
    "id": "vyLu2NkN70xkS",
    "kind": "multiple_choice",
    "properties": Object {
      "allowMultiple": true,
      "choices": Array [
        Object {
          "id": "F7XrAjByBk2w",
          "label": "Group Choice 1",
        },
        Object {
          "id": "cFEkKnnXmBP6",
          "label": "Group Choice 2",
        },
      ],
    },
    "title": "Group Multiple Choose ",
    "validations": Object {
      "required": true,
    },
  },
  Object {
    "id": "1rc8Rgj9et52",
    "kind": "text",
    "properties": Object {
      "choices": Array [],
    },
    "title": "üåü Your Name ",
    "validations": Object {
      "required": false,
    },
  },
  Object {
    "id": "bcFvi1syrp66",
    "kind": "phone_number",
    "properties": Object {
      "choices": Array [],
      "defaultCountry": "US",
    },
    "title": "‚òéÔ∏è  Your MobilePhone ",
    "validations": Object {
      "required": false,
    },
  },
  Object {
    "id": "yLxkSvkN7u2N",
    "kind": "multiple_choice",
    "properties": Object {
      "allowMultiple": true,
      "choices": Array [
        Object {
          "id": "kF7XrAjByB2w",
          "label": "Choice 1",
        },
        Object {
          "id": "BP6cFEkKnnXm",
          "label": "Choice 2",
        },
      ],
    },
    "title": "üß± Multiple Choose ",
    "validations": Object {
      "required": true,
    },
  },
  Object {
    "id": "MVlZfnmjZlye",
    "kind": "multiple_choice",
    "properties": Object {
      "allowMultiple": false,
      "choices": Array [
        Object {
          "id": "14wZu5EyC73q",
          "label": "üç´ Chocolate",
        },
        Object {
          "id": "MesZoaRCdR5Y",
          "label": "üçç Pineapple",
        },
      ],
    },
    "title": "üëã Just choose one",
    "validations": Object {
      "required": true,
    },
  },
  Object {
    "id": "X9F47wrnJImk",
    "kind": "yes_no",
    "properties": Object {
      "choices": Array [
        Object {
          "id": "adUm2Hhr1OOW",
          "label": "Yes",
        },
        Object {
          "id": "QMR1nZV0u84M",
          "label": "No",
        },
      ],
    },
    "title": "üëç Do you agree with us ",
    "validations": Object {
      "required": true,
    },
  },
  Object {
    "id": "EF9tbsJj3CNZ",
    "kind": "picture_choice",
    "properties": Object {
      "allowMultiple": false,
      "choices": Array [
        Object {
          "id": "XKV4vls1pkAX",
          "image": "http://qej6l0046.bkt.clouddn.com/form/5f2b64476481e3096526715b/1596679926__0-I_F2_OBINdzLx39o3nvZfzXwepkYO9mey0__doge.jpeg",
          "label": "Choice 1",
        },
        Object {
          "id": "FN5cCteg6SOa",
          "image": "http://qej6l0046.bkt.clouddn.com/form/5f2b64476481e3096526715b/1596679934__iAGYraz-j8J7gIxaC8GDJi017YJWUS16EzU7__p640.jpg",
          "label": "Choice 2",
        },
        Object {
          "id": "1Fl7MP9Nv2TM",
          "image": "http://qej6l0046.bkt.clouddn.com/form/5f2b64476481e3096526715b/1596679940__IKke45HeAlbu-v_yqOvbUjQHffRQRFSdl8OJ__p225-2.jpeg",
          "label": "Choice 3",
        },
      ],
    },
    "title": "üê∂ Choose the dog ",
    "validations": Object {
      "required": true,
    },
  },
  Object {
    "id": "iiOqCptRe2SN",
    "kind": "paragraph",
    "properties": Object {
      "choices": Array [],
    },
    "title": "üåπ What else do you want to say ",
    "validations": Object {
      "required": false,
    },
  },
  Object {
    "id": "Ri2SiOqCpteN",
    "kind": "code_block",
    "title": "üÄÑÔ∏è Do some code ",
    "validations": Object {
      "required": false,
    },
  },
  Object {
    "id": "qCpteNRi2SiO",
    "kind": "file_upload",
    "title": "üà¥Ô∏è Upload you avatar ",
    "validations": Object {
      "required": false,
    },
  },
  Object {
    "id": "NqCpSqCpteiO",
    "kind": "file_upload",
    "title": "You sig here",
    "validations": Object {
      "required": false,
    },
  },
  Object {
    "id": "qFN5cteg6SOa",
    "kind": "rating",
    "properties": Object {
      "total": 10,
    },
    "title": "ü™ê Rate here ",
    "validations": Object {
      "required": false,
    },
  },
  Object {
    "id": "g6Oayu81jkdqcte",
    "kind": "full_name",
    "title": "Full name",
    "validations": Object {
      "required": false,
    },
  },
  Object {
    "id": "NteSqF9g5cOa6",
    "kind": "input_table",
    "properties": Object {
      "tableColumns": Array [
        Object {
          "id": "name",
          "title": "Name",
        },
        Object {
          "id": "age",
          "title": "Age",
        },
      ],
    },
    "title": "Input table",
    "validations": Object {
      "required": true,
    },
  },
  Object {
    "id": "u1NyLxkSvkN7",
    "kind": "multiple_choice",
    "properties": Object {
      "allowMultiple": true,
      "allowOther": true,
      "choices": Array [
        Object {
          "id": "krAjByF7XB2w",
          "label": "Choice 1",
        },
        Object {
          "id": "P6cnBFEkKnXm",
          "label": "Choice 2",
        },
      ],
    },
    "title": "‚úà Multiple Choose with other",
    "validations": Object {
      "required": true,
    },
  },
  Object {
    "id": "Svky0NLxkN7u",
    "kind": "multiple_choice",
    "properties": Object {
      "allowMultiple": false,
      "allowOther": true,
      "choices": Array [
        Object {
          "id": "kB2wF7XrAjBy",
          "label": "Choice 1",
        },
        Object {
          "id": "BPnXm6cFEkKn",
          "label": "Choice 2",
        },
      ],
    },
    "title": "üéâ Single Choose with other",
    "validations": Object {
      "required": true,
    },
  },
]
`;
